chore(gatsby): convert babel-loader-helpers to typescript

Kornil opened the pull request:
Convert babel-loader-helpers to TS.

On line `101` and `109`, I get a type error that I am not able to fix related to `@babel/core` types not allowing `createConfigItem` to have a `name` key in its payload.

I'm not sure why tests are failing, other than the imports, the types shouldn't have affected prod. Any insights on what is causing it? Switching back the `babel-loader` imports doesn't do anything.

## Related Issues

https://github.com/gatsbyjs/gatsby/issues/21995

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -16,13 +29,23 @@ const loadCachedConfig = () => {
 }
 
-const getCustomOptions = stage => {
+export const getCustomOptions = (stage: string): IBabelStage["options"] => {
```
There's a string enum `Stage` which should be used for the argument here. You'll find it in commands/types.ts

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -98,17 +121,34 @@ const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
   const index = _.findIndex(
     items,
-    i => i.file.resolved === itemToMerge.file.resolved
+    i => i.file!.resolved === itemToMerge.file!.resolved
```
Can you use optional chaining rather than not-null assertion here and elsewhere, so avoid potential runtime errors

Kornil commented:
Thank you @ascorbic for the review. I updated the test types, the snapshots, and used the `Stage` enum instead of the `string` I had before. I also took the opportunity to clean up the `babelrc` reducer types since they had a duplicate stage key (introduced by me in a previous PR).

I added a :+1: on your comments to indicate I solved the issues. If you feel the feedback was addressed properly, feel free to resolve the conversations!

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -70,17 +94,17 @@ const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
+  pluginBabelConfig.stages[stage!].plugins.forEach(plugin => {
     reduxPlugins.push(
       babel.createConfigItem([resolve(plugin.name), plugin.options], {
         name: plugin.name,
```
Looks like this is an error. It seems it's correct that `name` isn't a valid key. It accepts just `dirname` and `type`: https://babeljs.io/docs/en/babel-core#createconfigitem

Kornil replied:
Wat do you think would be the best option here? I'm not sure how it will affect prod to remove the `name` key.

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -16,13 +30,23 @@ const loadCachedConfig = () => {
 
-const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
-  let pluginBabelConfig = loadCachedConfig()
+type prepareOptionsType = (
```
Rather than typing the whole function like this, can you just type the options object?

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -70,17 +94,17 @@ const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
-  pluginBabelConfig.stages[stage].plugins.forEach(plugin => {
+  const reduxPlugins: PluginItem[] = []
+  const reduxPresets: PluginItem[] = []
+  pluginBabelConfig.stages[stage!].plugins.forEach(plugin => {
```
If you've typed the options, and the stage key as optional you'll need a runtime null check for `stage` rather than this assertion.

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -70,17 +94,17 @@ const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
     )
   })
-  pluginBabelConfig.stages[stage].presets.forEach(preset => {
+  pluginBabelConfig.stages[stage!].presets.forEach(preset => {
```
Same. If you've done a runtime check then this will have been narrowed to exclude null.

ascorbic commented on packages/gatsby/src/utils/babel-loader-helpers.ts:
```diff
@@ -98,17 +122,34 @@ const prepareOptions = (babel, options = {}, resolve = require.resolve) => {
 }
 
-const mergeConfigItemOptions = ({ items, itemToMerge, type, babel }) => {
+type mergeConfigItemOptionsType = ({
```
As above, rather than typing the whole function like this, type the props object.

Kornil commented:
Addressed the second wave of feedback, let me know your opinion about the `name` issue on line `98` and `106`.

j0sh77 commented:
Hey! Thanks so much for opening this pull request!

Sadly this PR got stale and didn't have any activity for some time. We're trying to do better with PR reviews! To get a better overview of all actionable PRs we're going through all open PRs and triage them. Since we won't be able to do everything and adding new features _always_ means added maintenance burden, we have to be more picky about what's beneficial for the average user and the project itself longterm.

We think this is a great PR and would love to see it land in Gatsby. We're closing this PR _for now_ but if you're able to rebase onto the latest `master` branch and try it out with the latest `next` versions, we'd be happy to review the PR when its re-opened.

We absolutely want to have you as a contributor and are sorry for any inconveniences we caused with replying too late to this PR.

Thanks for submitting to Gatsby! :purple_heart:
